### PR TITLE
Add IF EXISTS to DROP ROLE

### PR DESCRIPTION
**Pull request by dekimir**
## Description

Add optional `IF EXISTS` to `DROP ROLE`, which suppresses the role-not-found error that you'd otherwise get.

## Additional context and related issues

Although this option isn't in the ANSI SQL, it's a UX improvement that Trino has traditionally admitted.

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# Section
* Extend `DROP ROLE` to allow `IF EXISTS`.
```

### CODE REVIEW

File: core/trino-parser/src/main/java/io/trino/sql/tree/DropRole.java [hunk 1]
```diff
@@ -55,6 +57,11 @@ public Optional<Identifier> getCatalog()
         return catalog;
     }
 
+    public boolean isIfExists()
```
Comment (martint) [inline core/trino-parser/src/main/java/io/trino/sql/tree/DropRole.java:60]: `isExists`, for consistency with all the other classes (DropTable, DropView, DropSchema, etc)

File: core/trino-main/src/test/java/io/trino/execution/BaseRoleTaskTest.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+import static java.util.concurrent.Executors.newCachedThreadPool;
+import static org.junit.jupiter.api.TestInstance.Lifecycle.PER_CLASS;
+import static org.junit.jupiter.api.parallel.ExecutionMode.CONCURRENT;
+
+/**
+ * Base class for various *RoleTask test classes.  Distinct from {@link BaseDataDefinitionTaskTest} due to different catalogs and metadata.
+ */
+@TestInstance(PER_CLASS)
+@Execution(CONCURRENT)
+public class BaseRoleTaskTest
```
Comment (martint) [inline core/trino-main/src/test/java/io/trino/execution/BaseRoleTaskTest.java:57]: Don't use inheritance for this. It makes it harder to maintain the tests, as the subclasses are intimately coupled to the details of the parent. Over time, as some of the tests evolve, the parent becomes a kitchen sink of functionality.

**Review by martint — CHANGES_REQUESTED** [review]

File: core/trino-parser/src/main/java/io/trino/sql/tree/DropRole.java (hunk 1 above)
Comment (dekimir) [reply core/trino-parser/src/main/java/io/trino/sql/tree/DropRole.java]: Done.

File: core/trino-main/src/test/java/io/trino/execution/BaseRoleTaskTest.java (hunk 2 above)
Comment (dekimir) [reply core/trino-main/src/test/java/io/trino/execution/BaseRoleTaskTest.java:57]: What would be a good way to factor out the boilerplate and make new tests easy to add?  I thought about putting the DropRole tests and SetRole tests into the same class, but I like the regularity of `Test<class name>Task` naming.

I could, of course, copy-paste the boilerplate into `TestDropRoleTask`, but that felt unsatisfying.

File: docs/src/main/sphinx/sql/drop-role.md [hunk 3]
```diff
@@ -3,7 +3,7 @@
 ## Synopsis
 
 ```text
-DROP ROLE role_name
+DROP ROLE [ IF EXISTS ] role_name
```
Comment (mosabua) [inline docs/src/main/sphinx/sql/drop-role.md:6]: Add docs for this in the description as well

Comment (dekimir) [reply docs/src/main/sphinx/sql/drop-role.md:6]: Done.

File: core/trino-main/src/test/java/io/trino/execution/BaseRoleTaskTest.java (hunk 2 above)
Comment (martint) [reply core/trino-main/src/test/java/io/trino/execution/BaseRoleTaskTest.java:57]: If all the role tests require similar setup, it's an ok compromise to have them in the same class (e.g., TestRoleTasks). They may require different setup and boilerplate, though, at which point they'd need to get split into separate classes with their relevant setup code.

Comment (dekimir) [reply core/trino-main/src/test/java/io/trino/execution/BaseRoleTaskTest.java]: > it's an ok compromise to have them in the same class (e.g., TestRoleTasks).

Done.

> They may require different setup and boilerplate, though, at which point they'd need to get split into separate classes with their relevant setup code.

ACK

**Comment by dekimir** [comment]
The [CI failure](https://github.com/trinodb/trino/actions/runs/10077096657/job/27860563940?pr=21985#step:5:386) seems related to #21032.

**Comment by wendigo** [comment]
I've restarted cassandra job

File: core/trino-main/src/main/java/io/trino/execution/DropRoleTask.java [hunk 4]
```diff
@@ -61,6 +61,9 @@ public ListenableFuture<Void> execute(
         Session session = stateMachine.getSession();
         Optional<String> catalog = processRoleCommandCatalog(metadata, session, statement, statement.getCatalog().map(Identifier::getValue));
         String role = statement.getName().getValue().toLowerCase(ENGLISH);
+        if (!metadata.roleExists(session, role, catalog) && statement.isExists()) {
```
Comment (wendigo) [inline core/trino-main/src/main/java/io/trino/execution/DropRoleTask.java:64]: I'd move `statement.isExists` as first operand (it's constant so cheaper if false and doesn't resolve metadata call).

**Review by wendigo — APPROVED** [review]
LGTM % order of operands in the if condition

File: core/trino-main/src/main/java/io/trino/execution/DropRoleTask.java (hunk 4 above)
Comment (dekimir) [reply core/trino-main/src/main/java/io/trino/execution/DropRoleTask.java]: Done.

**Pull request merged**